app/master/cluster_master.py

```diff
@@ -143,6 +144,15 @@ def handle_slave_state_update(self, slave, new_slave_state):
         do_transition = slave_transition_functions.get(new_slave_state)
         do_transition(slave)
 
+    def _shutdown_slave(self, slave):
+        """
+        Puts slave in shutdown mode so it cannot receive new builds. The slave will be killed when finished with any
+        running builds.
+        :type slave: Slave
+        """
+        slave.set_shutdown_mode()
+        self._logger.info('Slave on {} was put in shutdown mode. (id: {})', slave.url, slave.id)
```

Comment:
Do we also want to remove the slave from the `_all_slaves_by_url` dict? This is what the API uses to display a list of slaves, so I feel like we might just want to remove it from the API's list of slaves when we've explicitly told a slave to shut down. (E.g., it should just disappear from the slave monitor dashboard, not show up as red?)